### PR TITLE
Fix deselecting a first subcategory in Pref->Mantid->Options unticks main category 

### LINKED ISSUE #16871: Switching default instruments changes displayed algorithm categories
Reported by Spencer Howells at ISIS

### Expected behavior

Changing the default instrument in the MantidPlot "View->Preferences" menu should not affect the algorithm categories that are displayed in the Algorithms pane.

### Actual behavior

If a user has deselected some sub-categories in "View->Preferences->Mantid->Options" and changes the default instrument the parent category disappears from the algorithm view.

### Steps to reproduce the behavior

1. Clear (or comment out) any `algorithms.categories.hidden` in the `Mantid.user.properties` files
1. Go to the `View->Preferences` dialog and select the `Mantid->Options` tab
1. Expand the `Workflow` category and deselect all but `MIDAS` and click OK. Confirm that `MIDAS` is the only category visible in the `Algorithms` pane
1. Go back to the `View->Preferences` dialog and on the `Mantid->Instrument` tab select a different instrument and click OK
1. Observe that the `Workflow` category is now missing from the `Algorithms pane`  

### Platforms affected

All

### PR DESCRIPTION
**Pull request by DavidFair**
Description of work.
Certain categories under "View->Preferences->Mantid->Options" would take the value of their first subcategory depending on if they themselves were a category. 

When we get the list of categories and subcategories we now recurse backwards checking for any parent categories which were first seen from a subcategory and default them to a tick on as they do not have a visibility property associated.

**To test:**
Open "View->Preferences->Mantid->Options" all categories and subcategories should render correctly.
Categories such as "Algorithm" will preserve the tick option as it is standalone. The "Crystal" category is derived from its sub categories so will always appear ticked. 
Ensure the users selection is preserved across applying and re-opening the settings window.

Fixes #16871 .

Release notes have been updated

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by mantid-roman** [comment]
The PR doesn't seem to fix issue #16871. Is it possible that it is fixing a different issue?

**Comment by DavidFair** [comment]
@mantid-roman On my end it seems to be resolving the issue - its a bit confusing as the cause of the bug isn't related to default instruments unlike what the ticket says.

Previously if you deselected the first subcategory for Crystal or Workflow for example it would untick those categories when you next loaded the options window. 

The steps to reproduce are:
1. Currently if you untick Alignment under Workflow. 
2. Apply or OK
3. Open options window again and notice Workflow is now unticked regardless of its subcategories.
4. If the user hits apply the category Workflow now is hidden as it was unticked on load.

Hopefully with this PR it preserves that tick resolving the bug however if you can find a test where it doesn't I would be glad to look further into it.

**Comment by mantid-roman** [comment]
The expected behaviour is that changing the default instrument doesn't change the displayed categories. And this is what I am testing.

1. Open Options.
2. Untick all subcategories of Workflow except MIDAS.
3. Click OK.
4. The algorithm pane shows only MIDAS under Workflow.
4. Change the default instrument.
5. Open Options again. Now both MIDAS and MLZ are visible.
6. The pane also shows them both now.

**Comment by DavidFair** [comment]
Ahhh I was looking for algorithms becoming hidden not visible. I'll look further into this.

**Comment by DavidFair** [comment]
@mantid-roman I've gone back and tested it and can confirm this happens. Unfortunately due to the nature of the bug this isn't easily resolved. 

MLZ is found from its subcategories loaded in. So the GUI loading code first sees `MLZ//DNS` instead of `MLZ` then `MLZ//DNS`.  Because of this `MLZ` doesn't have its own attributes for if it should be checked so we need to default to something which this code does. 

I originally consulted @martyngigg for what the tick should default to and we decided that if a subcategory exists we make the parent category ticked. This is the behaviour we are seeing with `MLZ` as there are subcategories associated with it however I am happy to change the behaviour if there are better approaches?

**Comment by mantid-roman** [comment]
@martyngigg @DavidFair When you untick a root category (Workflow for example) it disappears from the Algorithms pane. But when you open the options you see that Workflow is ticked back but still invisible in the pane. If you hit Cancel Workflow category is ticked but invisible which is confusing. If you hit OK/Apply instead the category will appear without your explicitly ticking it which is unexpected.

**Comment by martyngigg** [comment]
@DavidFair We discussed seeing if Qt support a "partial" tick status, i.e grayed out to suggest that not all is ticked but some. Did you look at that?

**Comment by DavidFair** [comment]
I've implemented partial ticks within the last commit which I completely forgot about. Due to the many edge cases involved (explained below) with these categories the best solution was to walk the widget tree after it is generated setting the parent categories based on their children's status. 
This should make the behaviour a bit more consistent however AFAIK there isn't any methods called when a user ticks or unticks a box so the partial ticks can only be generated when the options window is refreshed. If there is one I am missing its just as simple as adding a few lines of code to it so I am still looking for it. 

@mantid-roman Unfortunately Workflow does not have its own category, instead it is derived from its child categories which are returned from the `AlgorithmFactory`. Because of this Workflow has to figure out from these sub-categories whether to be ticked, unticked or partially ticked. 
If any subcategories of workflow are on then workflow will take the checked or partial option on loading the options windows which makes it reappear when hitting apply as you describe. 
Without Workflow being returned from the `AlgorithmFactory` as a stand-alone category with a state of on or off I can't save the tick status hence the workarounds in these commits in the mean time.

**Comment by mantid-roman** [comment]
In this case the check box of Workflow (and similar entries) should be disabled because editing them manually only adds confusion.

**Comment by martyngigg** [comment]
@DavidFair What about signals such as http://doc.qt.io/qt-4.8/qtreewidget.html#itemClicked, are they not of any use?

File: MantidPlot/src/ConfigDialog.cpp [hunk 1]
```diff
@@ -1192,10 +1192,14 @@ void ConfigDialog::updateProgramTree() {
     // requires them.
     std::map<std::string, std::string> programKeysAndDetails = itr->second;
 
+    Qt::CheckState checkStatus = Qt::Unchecked;
+    if (programKeysAndDetails.find("visible")->second == "Yes") {
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: Can `find` return an invalid iterator?

File: MantidPlot/src/ConfigDialog.cpp [hunk 2]
```diff
@@ -1267,59 +1271,176 @@ void ConfigDialog::updateSendToTab() {
...
+}
+
+void ConfigDialog::buildTreeCategoryStructure(
+    const QString *catNames, const bool isHidden,
+    QMap<QString, QTreeWidgetItem *> *seenCategories) {
+  /* Some categories come as standalone and some are discovered
+  // via their subcategory. E.g. Arithmetic vs Crystal\\Cell
+  // If standalone it has its own isHidden attribute otherwise
+  // we have to assume it to be visible (or ticked) as that information
+  // is not stored. */
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: Use only `//` for comments

File: MantidPlot/src/ConfigDialog.cpp [hunk 3]
```diff
@@ -1267,59 +1271,176 @@ void ConfigDialog::updateSendToTab() {
...
+  for (auto i = categoryMap.cbegin(); i != categoryMap.cend(); ++i) {
+
+    QString catNames = QString::fromStdString(i->first);
+    // Start recursion down building tree from names
+    buildTreeCategoryStructure(&catNames, i->second, &categories);
+  }
+}
+
+void ConfigDialog::buildTreeCategoryStructure(
+    const QString *catNames, const bool isHidden,
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: Can `catNames` be a `nullptr`? Use a reference if not.

File: MantidPlot/src/ConfigDialog.cpp [hunk 4]
```diff
@@ -1267,59 +1271,176 @@ void ConfigDialog::updateSendToTab() {
...
+    QMap<QString, QTreeWidgetItem *> *seenCategories) {
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: `seenCategories` should probably be a reference too.

File: MantidPlot/src/ConfigDialog.cpp [hunk 5]
```diff
@@ -1267,59 +1271,176 @@ void ConfigDialog::updateSendToTab() {
...
+
+  // Split the categories whose name is delimited by '\\'
+  QStringList splitCats = catNames->split('\\');
+
+  // Create a unique key for storing into the mapping
+  QString catKey(*catNames);
+  // We need to remove delimiter as top level categories do not come with it
+  catKey.remove('\\');
+
+  int splitCatIndex = splitCats.size();
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: Can `splitCatIndex` be 0?

File: MantidPlot/src/ConfigDialog.cpp [hunk 6]
```diff
@@ -1267,59 +1271,176 @@ void ConfigDialog::updateSendToTab() {
...
+  if (splitCatIndex == 0) {
+    // This is a top level category and standalone
+    QTreeWidgetItem *catWidget =
+        createCheckedTreeItem(splitCats[0], tickStatus);
+    treeCategories->addTopLevelItem(catWidget);
+    seenCategories->insert(catKey, catWidget);
+  } else {
+    // We need to determine if we have seen this category before
+    // and get any pointers to its parent(s) and set their tick status
+    int prevElement = splitCatIndex - 1;
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: Let's hope that `splitCatIndex` isn't 0.

File: MantidPlot/src/ConfigDialog.cpp [hunk 7]
```diff
@@ -1267,59 +1271,176 @@ void ConfigDialog::updateSendToTab() {
...
+  // We have built tree now fix any ticks or make partial ticks
+  int topLevelCats = treeCategories->topLevelItemCount();
+  for (int i = 0; i < topLevelCats; i++) {
+    correctTreePatrialTicks(treeCategories->topLevelItem(i));
+  }
+}
+
+QTreeWidgetItem *ConfigDialog::walkBackwardsThroughCategories(
+    const QString *catNames, int elementToCheck,
+    QMap<QString, QTreeWidgetItem *> *seenCategories,
```
Comment (mantid-roman) [inline MantidPlot/src/ConfigDialog.cpp]: References?

**Comment by DavidFair** [comment]
@mantid-roman I've gone through and changed the parameters to use references and also converted the loops to for ranged loops. Also the tick box behaviour should be more consistent throughout now. 
Would you be able to give it another once over to see if there is anything else that could be improved?

**Comment by mantid-roman** [comment]
@DavidFair No problem

**Comment by mantid-roman** [comment]
The Options and the Algorithms pane seem consistent now. One more little thing to consider. There is no way to hide all sub-categories and keep only the top level algorithms. For example, if you untick Diffraction\UsesPropertyManger all Diffraction disappears.

**Comment by DavidFair** [comment]
Ok following a discussion with Roman I am happy with it in its current state. 
If unticking subcategories such as UsesPropertyManager causes issue for users we can look into modifying the `AlgorithmFactory` however that is out of the scope of this ticket.

**Comment by mantid-roman** [comment]
I am happy to :shipit:

File: MantidPlot/src/ConfigDialog.cpp [hunk 8]
```diff
@@ -715,8 +715,8 @@ void ConfigDialog::initMantidPage() {
 
   // Populate boxes
   auto faclist = cfgSvc.getFacilityNames();
-  for (auto it = faclist.begin(); it != faclist.end(); ++it) {
-    facility->addItem(QString::fromStdString(*it));
+  for (auto it : faclist) {
```
Comment (FedeMPouzols) [inline MantidPlot/src/ConfigDialog.cpp]: In this and other loops below you probably can use `const auto &`. It's not going to do any difference here, but I'd say it's good practice in general to try to use all those specifiers whenever possible.

**Pull request merged**